src/Option/Option.ts

```diff
@@ -108,7 +108,7 @@ const getMatcherResult = <T, R>(match: ((t: T) => R) | R, arg: T) =>
  * ) // => 84
  */
 export const match =
-    <A extends NonNullish, R>(matcher: OptionMatcher<A, R>) =>
+    <A extends NonNullish, R>(matcher: OptionMatcher<NoInfer<A>, R>) =>
```

Comment:
the gist of all these changes is that we would prefer TypeScript to infer the types of `A` (and `E` in the case of Result) from the option/result instance passed in the pipeline, and **not** from the curried function.